### PR TITLE
Use HMAC authentication for network packets - 2nd attempt

### PR DESCRIPTION
**Pull request by SimonWalton-HiFi**
This changes the network authentication scheme to use HMAC-MD5 (from the openssl lib) rather than just MD5. This is based on earlier PR12690.

https://github.com/highfidelity/hifi/pull/12690

This version fixes a race condition that could occur because sending of reliable packets is performed on a different thread to the NodeList thread used for all other traffic.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9188/app_13005_PR_66bd424ae4002a75cd3003f4e97df1322dd2143f.apk

**Comment by SimonWalton-HiFi** [comment]
## Test Plan

QA must run a local domain to connect. Any connectivity to existing clients or server should result in denial due to protocol mismatch.

Connectivity to clients or server that have this PR should work normally with no noticeable change in performance. Examine all client and domain server logs for errors of the form "Packet hash mismatch ...".
They must not appear.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9188/HighFidelity-Beta-PR13005-66bd424ae4002a75cd3003f4e97df1322dd2143f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9188/HighFidelity-Beta-Interface-PR13005-66bd424ae4002a75cd3003f4e97df1322dd2143f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9188/HighFidelity-Beta-PR13005-66bd424ae4002a75cd3003f4e97df1322dd2143f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9188/HighFidelity-Beta-Interface-PR13005-66bd424ae4002a75cd3003f4e97df1322dd2143f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9196/HighFidelity-Beta-PR13005-9ef56c44a3da38161741cf4dc62f49e786c8e5b1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9196/HighFidelity-Beta-Interface-PR13005-9ef56c44a3da38161741cf4dc62f49e786c8e5b1.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9196/HighFidelity-Beta-PR13005-9ef56c44a3da38161741cf4dc62f49e786c8e5b1.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9196/HighFidelity-Beta-Interface-PR13005-9ef56c44a3da38161741cf4dc62f49e786c8e5b1.exe

**Comment by SimonWalton-HiFi** [comment]
I put in some timing code for the hash calculation. Typical times are 2-5 us. Occasionally a longer period up to 17 us, but fairly rare.

File: libraries/networking/src/HMACAuth.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    hashResult.resize(EVP_MAX_MD_SIZE);
+    unsigned int hashLen;
+    if (HMAC_Final(_hmacContext, &hashResult[0], &hashLen)) {
+        hashResult.resize((size_t)hashLen);
+        // Clear state for possible reuse.
+        HMAC_Init_ex(_hmacContext, nullptr, 0, nullptr, nullptr);
+        return true;
+    }
```
Comment (birarda) [inline libraries/networking/src/HMACAuth.cpp]: Could we clean this up a bit by making the QMutex recursive and calling addData and then result() from ere?

Comment (SimonWalton-HiFi) [reply libraries/networking/src/HMACAuth.cpp]: Changed this now.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9526/app_13005_PR_00d2f9494bbc493a8efad13ea2cbe29a1983339c.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9526/HighFidelity-Beta-PR13005-00d2f9494bbc493a8efad13ea2cbe29a1983339c.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9526/HighFidelity-Beta-Interface-PR13005-00d2f9494bbc493a8efad13ea2cbe29a1983339c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9526/HighFidelity-Beta-PR13005-00d2f9494bbc493a8efad13ea2cbe29a1983339c.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9526/HighFidelity-Beta-Interface-PR13005-00d2f9494bbc493a8efad13ea2cbe29a1983339c.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9612/app_13005_PR_0349847857fb2393c0894426b22011a60c2475a5.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9612/HighFidelity-Beta-PR13005-0349847857fb2393c0894426b22011a60c2475a5.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9612/HighFidelity-Beta-Interface-PR13005-0349847857fb2393c0894426b22011a60c2475a5.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9612/HighFidelity-Beta-PR13005-0349847857fb2393c0894426b22011a60c2475a5.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9612/HighFidelity-Beta-Interface-PR13005-0349847857fb2393c0894426b22011a60c2475a5.exe

**Comment by SimonWalton-HiFi** [comment]
Version bumped.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9649/app_13005_PR_9b6225db40bfa1f69707e2a1507a007f2eed9b4f.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9649/HighFidelity-Beta-PR13005-9b6225db40bfa1f69707e2a1507a007f2eed9b4f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9649/HighFidelity-Beta-Interface-PR13005-9b6225db40bfa1f69707e2a1507a007f2eed9b4f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9649/HighFidelity-Beta-PR13005-9b6225db40bfa1f69707e2a1507a007f2eed9b4f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9649/HighFidelity-Beta-Interface-PR13005-9b6225db40bfa1f69707e2a1507a007f2eed9b4f.exe

**Comment by birarda** [comment]
#### Test Plan

Smoke test Interface connecting to the Sandbox from this PR, with multiple users. Ensure that you hear audio, see the other avatar, and can see entities created by the other user.

Ensure that the Interface, domain-server, and assignment-client do not contain any messages that begin with: "Packet hash mismatch on"

**Comment by qligmaloney** [comment]
Currently testing

**Pull request merged**